src/typescript-worker/node/resolve.ts

```diff
@@ -8,144 +8,179 @@
 // https://github.com/nodejs/node/blob/a9dd03b1ec89a75186f05967fc76ec0704050c36/lib/internal/modules/esm/resolve.js
 // and adapted for use in playground-elements.
 
-/**
- * @param {string} match
- * @param {URL} pjsonUrl
- * @param {boolean} isExports
- * @param {string | URL | undefined} base
- * @returns {void}
- */
-function emitFolderMapDeprecation(match, pjsonUrl, isExports, base) {
-  const pjsonPath = fileURLToPath(pjsonUrl);
+import {fileURLToPath} from './url.js';
+import {
+  InvalidModuleSpecifierError,
+  InvalidPackageConfigError,
+  InvalidPackageTargetError,
+  PackagePathNotExportedError,
+} from './errors.js';
+import type {
+  PackageExports,
+  PackageExportsPathOrConditionMap,
+  PackageExportsTarget,
+  PackageJsonWithExports,
+} from '../util.js';
 
-  if (emittedPackageWarnings.has(pjsonPath + '|' + match)) return;
-  emittedPackageWarnings.add(pjsonPath + '|' + match);
-  process.emitWarning(
+function emitFolderMapDeprecation(
+  match: string,
+  pjsonUrl: URL,
+  isExports: boolean,
+  base: string
+): void {
+  const pjsonPath = fileURLToPath(pjsonUrl);
+  console.warn(
     `Use of deprecated folder mapping "${match}" in the ${
       isExports ? '"exports"' : '"imports"'
     } field module resolution of the package at ${pjsonPath}${
-      base ? ` imported from ${fileURLToPath(base)}` : ''
+      base ? ` imported from ${base}` : ''
     }.\n` +
       `Update this package.json to use a subpath pattern like "${match}*".`,
     'DeprecationWarning',
     'DEP0148'
   );
 }
 
-/**
- * @param {string} specifier
- * @param {URL} packageJSONUrl
- * @param {string | URL | undefined} base
- */
-function throwExportsNotFound(subpath, packageJSONUrl, base) {
-  throw new ERR_PACKAGE_PATH_NOT_EXPORTED(
+function makeExportsNotFoundError(
+  subpath: string,
+  packageJSONUrl: URL,
+  base: string
+): PackagePathNotExportedError {
+  throw new PackagePathNotExportedError(
     fileURLToPath(new URL('.', packageJSONUrl)),
     subpath,
-    base && fileURLToPath(base)
+    base
```

Comment:
Yeah,  in the original code `base` can be a URL, string, or undefined, presumably because of some different ways Node can invoke its resolve function. However in our case, we know we always have a string, so I just tightened the type to reflect that.